llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp

```diff
@@ -3914,8 +3914,12 @@ LegalizerHelper::lower(MachineInstr &MI, unsigned TypeIdx, LLT LowerHintTy) {
   case G_SSHLSAT:
   case G_USHLSAT:
     return lowerShlSat(MI);
-  case G_ABS:
+  case G_ABS: {
+    LLT Ty = MRI.getType(MI.getOperand(0).getReg());
+    if (Ty.isScalar() &&  LI.isLegalOrCustom({G_ICMP, {Ty, Ty}}))
+        return lowerAbsToCNeg(MI);
```

Comment:
I personally don't know if there is a good check that captures what we would want through the isLegalOrCustom interface. I think many targets would want the ashr/AddXor form even if they have legal sub+sel. I would remove this code and just keep the AArch64LegalizerInfo::legalizeCustom method.

Reply:
Reusing the same type for the compare result isn't right, but we don't have a great API for this. We kind of want the equivalent of getSetCCResultType. A more awkward solution would be to query with s1 and then look through the legalize actions until you find transforms on the original. Failing that, using i1 as the basis of the query is a more reasonable choice I think

Reply:
My main point is that we should do something akin to the DAG. We check for the legality of icmp and then select a select without checking whether it is legal or not. The icmp check is insufficient.

Reply:
@tschuett How would that handle @arsenm's concern?

I agree with @davemgreen's approach and will update the patch accordingly. I would not like to touch `lower` as of now and handle it in `legalizeCustom`. Given that we don't have the right APIs, the right approach would be handle this separately.

Reply:
These are independent issues. We have no API to query the result type from compares. But building a Select without checking whether it is legal to do so may cause trouble.

Reply:
but that would still reuse the same type (as @arsenm pointed out). Isn't it? So, how would it solve the core issue?
I understand the lack of API and thus I would like to avoid changing `lower()` in this patch.